source/extensions/network/dns_resolver/cares/dns_impl.cc

```diff
@@ -66,6 +69,30 @@ absl::optional<std::string> DnsResolverImpl::maybeBuildResolversCsv(
   return {absl::StrJoin(resolver_addrs, ",")};
 }
 
+std::vector<in_addr> DnsResolverImpl::maybeBuildResolversVector(
+    const std::vector<Network::Address::InstanceConstSharedPtr>& resolvers) {
+  if (resolvers.empty()) {
+    return {};
+  }
+
+  std::vector<in_addr> resolver_addrs;
+
+  for (const auto& resolver : resolvers) {
+    if (resolver->ip() == nullptr) {
+      throw EnvoyException(
+          fmt::format("DNS resolver '{}' is not an IP address", resolver->asString()));
+    }
+
+    // c-ares only supports ipv4 addresses set this way
+    if (resolver->ip()->ipv4()) {
```

Comment:
It seems odd to me that we fail above if it's not an IP address, but we ignore of v6? I think we should fail with v6 so it's more obvious?

Reply:
This comment plus the one [below](https://github.com/envoyproxy/envoy/pull/19010#discussion_r762092205) made me realize it is silly to keep both representations even though they are mutually exclusive. The solver will _either_ use the user defined resolvers as fallback _or_ as override. Cleaning this up to make it easier to follow.